test/metadata.jl

```diff
@@ -0,0 +1,3065 @@
+module TestMetadata
+
+using Test, DataFrames, Random
+
+@testset "table-level metadata" begin
+    for x in (DataFrame(), DataFrame(a=1))
+        @test isempty(metadatakeys(x))
+        @test metadatakeys(x) isa Tuple
+        @test getfield(x, :allnotemetadata)
+        metadata!(x, "name", "empty", style=:some)
+        @test !getfield(x, :allnotemetadata)
+        @test collect(metadatakeys(x)) == ["name"]
+        @test metadatakeys(x) isa Base.KeySet
+        @test metadata(x, "name") == "empty"
+        @test metadata(x, "name", style=true) == ("empty", :some)
+        emptymetadata!(x)
+        @test !getfield(x, :allnotemetadata)
```

Comment:
OK - I will add a separate set of tests. But in summary `:allnotemetadata` is touched in:
* `DataFrame` constructor, where it is set to `true` by default
* `copy` where it is copied from the source
* `metadata!`, where it is set to `false` if metadata is not `:note`-style
* `colmetadata!`, where it is set to `false` if metadata is not `:note`-style
* `_drop_all_nonnote_metadata!` where it is set to `true`

So in short it can become `true` again in an existing `DataFrame` only if `_drop_all_nonnote_metadata!` is called.

Reply:
I added a special function in tests and added tests everywhere.

Reply:
Ah OK, so the logic is less complex than I feared. But it's nice to have good testing anyway!

Reply:
I could make more complex logic, but it would be hard to maintain. The core idea is that if `_drop_all_nonnote_metadata!` is called repeatedly (it is called in many functions) then it should do removal of non-`:note` metadata only once.